src/SignalR/clients/java/signalr/src/main/java/com/microsoft/signalr/MessagePackHubProtocol.java

```diff
@@ -0,0 +1,580 @@
+                break;
+            case STRING:
+                item = unpacker.unpackString();
+                break;
+            case BINARY:
+                length = unpacker.unpackBinaryHeader();
+                byte[] binaryValue = new byte[length];
+                unpacker.readPayload(binaryValue);
+                item = binaryValue;
+                break;
+            case ARRAY:
+                length = unpacker.unpackArrayHeader();
+                Object[] arrayValue = new Object[length];
+                for (int i = 0; i < length; i++) {
+                    arrayValue[i] = readValue(unpacker, new Object().getClass());
+                }
+                item = arrayValue;
+                //If the itemType is an array, we return an array. Else we convert the array to a list.
+                if (!itemType.isArray()) {
+                    item = new ArrayList<Object>(Arrays.asList(arrayValue));
+                }
+                break;
+            case MAP:
+                length = unpacker.unpackMapHeader();
+                Map<Object, Object> mapValue = new HashMap<Object, Object>();
+                for (int i = 0; i < length; i++) {
+                    Object key = readValue(unpacker, new Object().getClass());
+                    Object value = readValue(unpacker, new Object().getClass());
+                    mapValue.put(key, value);
+                }
+                item = mapValue;
+                break;
+            case EXTENSION:
+                /*
+                ExtensionTypeHeader extension = unpacker.unpackExtensionTypeHeader();
+                byte[] extensionValue = new byte[extension.getLength()];
+                unpacker.readPayload(extensionValue);
+                //Convert this to an object?
+                item = extensionValue;
+                */
+                throw new RuntimeException("Extension types are not supported yet");
+            default:
+                break;
+        }
+        return itemType.cast(item);
+    }
+
+    private void writeValue(Object o, MessagePacker packer) throws IOException {
+
+        if (o == null) {
+            packer.packNil();
+        } else if (o instanceof Boolean) {
+            packer.packBoolean((boolean) o);
+        } else if (o instanceof BigInteger) {
+            packer.packBigInteger((BigInteger) o);
+        } else if (o instanceof Long) {
+            packer.packLong((long) o);
+        } else if (o instanceof Short) {
+            packer.packShort((short) o);
+        } else if (o instanceof Integer) {
+            packer.packInt((int) o);
+        } else if (o instanceof Double) {
+            packer.packDouble((double) o);
+        } else if (o instanceof Float) {
+            packer.packFloat((float) o);
+        } else if (o instanceof String) {
+            packer.packString((String) o);
+        } else if (o instanceof Byte) {
+            packer.packByte((byte) o);
+        // Unsure about this
```

Comment:
Similarly here, I'm assuming Collections can all be packed as arrays

Reply:
I think that makes sense. Does JsonHubProtocol turn all Collections into JSON arrays?